docs/contributing/git-style.rst

```diff
@@ -46,3 +46,98 @@ features or plugins to streamline this process (e.g.,
+
+- If this change fixes a bug, what was the original bug or its symptoms?
+
+- If you did some additional changes along the way to facilitate the main
+  thrust of your work, explain the rational behind any non-obvious but required
+  changes.
+
+At the same time, there is no reason to agonize over the content of a commit
+message or be overly verbose. Applying an automatic formatter changes a lot of
+files but doesn't require a lot of context, while a one-line change might merit
+several paragraphs of context if the rationale is subtle or non-obvious.
+
+The subject line (i.e., the first line) of a commit message should be short,
+generally less than 80 characters long, and starts with a verb in the
+imperative mood (present tense, second person). You can think of the message as
+directing the codebase itself to do something. Optionally, the subject line may
+begin with a component followed by a colon if the change is specific to a
+commonly known component or library.
+
+Examples:
+
+- ``Apply clang-format``
+
+- ``tsuba: Support compressed and uncompressed files``
+
+- ``support: Fix release order when handling errors``
+
+Commit messages are traditionally viewed in a terminal. Hard wrap your lines to
+some reasonable column width (e.g., 80 or 120 characters) and use a blank line
+to separate paragraphs. Avoid markup or excessive decoration on text. A commit
+message is plain text.
+
+Refrain from using external links unless you are sure they will be valid
+indefinitely. Assume that readers do not have access to external links and
+summarize relevant details in the commit message itself. This includes
+references to issue and work trackers.
+
+An example of a typical commit message:
+
+.. code-block::
+
+   support: Reduce memory usage of Foo
+
+   A Foo instance maintains references to bar objects. The number of bar
+   objects can grow when a Foo is used multiple times. This can reach 10GB in
+   typical usage.
+
+   In practice, once a Foo is reused only some bar objects need to be retained.
+   So, add a dispose API that allows Foo users to indicate which bar objects to
+   release on reuse. The default behavior is to retain all bars.
+
+   In most cases, it is easy to infer which bar objects to dispose of by code
+   inspection. The few remaining cases where it is not possible to use the
+   dispose API are on less frequently executed recovery paths.
+
+   After this change, typical memory usage reduces to a few KB.
+
+   JIRA: KAT-1000
+   GitHub: #2222
+
+At the end of the day, remember that there are three main things we accomplish
+with commit messages:
+
+1. Facilitate review of our changes by reviewers
+
+2. Provide context to future readers about our changes
+
+3. Ground the scope of our changes to ourselves. If we struggle to write a
+   simple explanation of why we made a change, it is a sign that the underlying
+   code changes are also not as clear as they could be.
```

Comment:
I think this should be at the top not the bottom, or maybe both. Best to give people an idea of the goals immediately to have them in the right mindset.